File: app/services/qbit.py
import httpx
import logging
import re
from datetime import datetime
from guessit import guessit
from sqlmodel import Session, select
from app.config import get_settings
from app.models import Game, GameStatus
from app.services.igdb import IGDBService
from app.utils import extract_version, sanitize_search_query

logger = logging.getLogger("repackarr")
settings = get_settings()


class QBitService:
    """
    Service for interacting with qBittorrent WebUI API.
    Syncs completed torrents from a configured category into the game library.
    """
    
    def __init__(self):
        self.base_url = settings.QBIT_HOST
        self.client = httpx.AsyncClient(timeout=30.0, verify=False)
        self.cookies: dict = {}
        self.igdb = IGDBService()

    async def login(self) -> bool:
        """
        Authenticate with qBittorrent WebUI.
        
        Returns:
            True if login successful, False otherwise
        """
        try:
            resp = await self.client.post(
                f"{self.base_url}/api/v2/auth/login",
                data={
                    "username": settings.QBIT_USERNAME,
                    "password": settings.QBIT_PASSWORD
                }
            )
            if resp.status_code == 200 and "SID" in resp.cookies:
                self.cookies = dict(resp.cookies)
                logger.info("Successfully authenticated with qBittorrent")
                return True
                
            logger.error(f"qBittorrent login failed: {resp.text}")
            return False
        except httpx.RequestError as e:
            logger.error(f"qBittorrent connection error: {e}")
            return False
        except Exception as e:
            logger.error(f"qBittorrent login error: {e}")
            return False

    async def sync_games(self, session: Session) -> int:
        """
        Sync games from qBittorrent to the database.
        
        Args:
            session: Database session
            
        Returns:
            Number of games synced/updated
        """
        if not self.cookies:
            if not await self.login():
                return 0

        try:
            resp = await self.client.get(
                f"{self.base_url}/api/v2/torrents/info",
                params={"category": settings.QBIT_CATEGORY},
                cookies=self.cookies
            )
            resp.raise_for_status()
            torrents = resp.json()
        except httpx.HTTPStatusError as e:
            logger.error(f"qBittorrent API error: {e.response.status_code}")
            return 0
        except httpx.RequestError as e:
            logger.error(f"Failed to fetch torrents: {e}")
            return 0
        except Exception as e:
            logger.error(f"Failed to fetch torrents: {e}")
            return 0

        synced_count = 0
        
        for torrent in torrents:
            try:
                if await self._process_torrent(torrent, session):
                    synced_count += 1
            except Exception as e:
                logger.error(f"Error processing torrent {torrent.get('name')}: {e}")
        
        session.commit()
        logger.info(f"Synced {synced_count} game(s) from qBittorrent")
        return synced_count

    async def _process_torrent(self, torrent: dict, session: Session) -> bool:
        """
        Process a single torrent and sync it to the database.
        
        Args:
            torrent: Torrent info from qBittorrent API
            session: Database session
            
        Returns:
            True if game was added/updated, False otherwise
        """
        # Get completion date, fallback to added date
        ts = torrent.get('completion_on', 0)
        if ts <= 0:
            ts = torrent.get('added_on', 0)
        
        if ts <= 0:
            return False
        
        # Use UTC datetime for consistency with Prowlarr/Database
        torrent_date = datetime.utcfromtimestamp(ts)
        
        raw_name = torrent.get('name', '')
        if not raw_name:
            return False
            
        logger.debug(f"Processing torrent: {raw_name}")
        
        # Parse title using guessit
        parsed = guessit(raw_name)
        title = parsed.get('title')
        
        if not title:
            logger.warning(f"Could not parse title from: {raw_name}")
            return False
        
        # Clean up the title - remove release group patterns
        # Remove known release groups (after -, ., or _)
        title = re.sub(r'[-._](?:CODEX|SKIDROW|RELOADED|CPY|FLT|PLAZA|RAZOR1911|HOODLUM|DOGE|RUNE|TiNYiSO|DARKSiDERS|ANOMALY|PROPHET|GOLDBERG|STEAMPUNKS|EMPRESS|DODI|FITGIRL|NECROS|ElAmigos|KaOs|GOG|TENOKE|P2P|insaneramzes)(?:[-._].*)?$', '', title, flags=re.IGNORECASE)
        # Remove trailing UPPERCASE-only words after - or . (safer - preserves "II", "DLC", "VR", lowercase groups)
        # This hybrid approach only removes all-caps release groups while keeping mixed/lower case words
        title = re.sub(r'[-._]([A-Z]{2,15})$', '', title)
        title = title.strip()
        
        if not title:
            logger.warning(f"Title became empty after cleanup from: {raw_name}")
            return False
        
        # Extract version
        detected_version = extract_version(raw_name)
        
        # Generate clean search query (removes versions, tags, special chars)
        search_query = sanitize_search_query(title)
        if not search_query:
            search_query = title  # Fallback to title if sanitization fails
        
        # Check if game already exists
        existing_game = session.exec(
            select(Game).where(Game.title == title)
        ).first()

        if not existing_game:
            return await self._add_new_game(
                title, search_query, detected_version, torrent_date, session
            )
        else:
            return await self._update_existing_game(
                existing_game, detected_version, torrent_date, session
            )

    async def _add_new_game(
        self,
        title: str,
        search_query: str,
        version: str | None,
        torrent_date: datetime,
        session: Session
    ) -> bool:
        """Add a new game to the database."""
        logger.info(f"New game detected: {title} (v{version or 'unknown'})")
        logger.debug(f"Search query: {search_query}")
        
        # Try to fetch metadata from IGDB
        cover_url = None
        steam_app_id = None
        if settings.is_igdb_enabled:
            try:
                metadata = await self.igdb.get_game_metadata(title)
                if metadata:
                    cover_url = metadata.get("cover_url")
                    steam_app_id = metadata.get("steam_app_id")
            except Exception as e:
                logger.debug(f"Failed to fetch metadata for {title}: {e}")
        
        new_game = Game(
            title=title,
            search_query=search_query,
            current_version_date=torrent_date,
            current_version=version,
            status=GameStatus.MONITORED,
            cover_url=cover_url,
            steam_app_id=steam_app_id
        )
        session.add(new_game)
        logger.info(f"Added {title} to library")
        return True

    async def _update_existing_game(
        self,
        game: Game,
        version: str | None,
        torrent_date: datetime,
        session: Session
    ) -> bool:
        """Update an existing game if newer version found."""
        updated = False
        
        # Update if torrent is newer
        if torrent_date > game.current_version_date.replace(tzinfo=None):
            logger.info(
                f"Updating {game.title}: "
                f"v{game.current_version or '?'} -> v{version or '?'}"
            )
            game.current_version_date = torrent_date
            if version:
                game.current_version = version
            session.add(game)
            updated = True
        
        # Retry metadata fetch if missing
        if (not game.cover_url or not game.steam_app_id) and settings.is_igdb_enabled:
            try:
                metadata = await self.igdb.get_game_metadata(game.title)
                if metadata:
                    if not game.cover_url and metadata.get("cover_url"):
                        game.cover_url = metadata["cover_url"]
                        updated = True
                    if not game.steam_app_id and metadata.get("steam_app_id"):
                        game.steam_app_id = metadata["steam_app_id"]
                        updated = True
                    if updated:
                        session.add(game)
            except Exception:
                pass
                
        return updated

    async def add_torrent(self, magnet_url: str, category: str = None) -> bool:
        """
        Add a torrent to qBittorrent via magnet link.
        
        Args:
            magnet_url: The magnet link
            category: Category to assign (defaults to settings.QBIT_CATEGORY)
            
        Returns:
            True if successful
        """
        if not self.cookies:
            if not await self.login():
                return False
                
        try:
            category = category or settings.QBIT_CATEGORY
            
            resp = await self.client.post(
                f"{self.base_url}/api/v2/torrents/add",
                data={
                    "urls": magnet_url,
                    "category": category,
                },
                cookies=self.cookies
            )
            
            if resp.status_code == 200:
                # "Ok." or similar text is returned on success, but 200 is main indicator
                logger.info("Sent magnet to qBittorrent")
                return True
            else:
                logger.error(f"Failed to add torrent: {resp.status_code} - {resp.text}")
                return False
                
        except Exception as e:
            logger.error(f"Exception adding torrent: {e}")
            return False
    
    async def close(self) -> None:
        """Close HTTP clients."""
        await self.client.aclose()
        await self.igdb.close()

File: src/lib/server/qbit.ts
import { settings, isIgdbEnabled } from './config.js';
import { db } from './database.js';
import { games } from './schema.js';
import { eq } from 'drizzle-orm';
import { logger } from './logger.js';
import {
	extractVersion,
	sanitizeSearchQuery,
	fuzzyMatchTitles,
	parseTorrentTitle,
	extractSourceUrl
} from './utils.js';
import { getGameMetadata } from './igdb.js';
import { retryAsync } from './validators.js';
import type { TorrentInfo } from './types.js';

class QBitService {
	private baseUrl: string;
	private cookies: string = '';

	constructor() {
		this.baseUrl = settings.QBIT_HOST;
	}

	async login(forceRefresh = false): Promise<boolean> {
		if (!this.baseUrl) {
			logger.warn('[qBit] QBIT_HOST is not configured — check your .env file.');
			return false;
		}
		if (forceRefresh) this.cookies = '';
		if (this.cookies) return true;
		try {
			const success = await retryAsync(
				async () => {
					const resp = await fetch(`${this.baseUrl}/api/v2/auth/login`, {
						method: 'POST',
						headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
						body: new URLSearchParams({ username: settings.QBIT_USERNAME, password: settings.QBIT_PASSWORD })
					});
					if (!resp.ok) return false;
					const text = await resp.text();
					if (resp.status !== 204 && text.trim() !== 'Ok.') return false;
					const setCookie = resp.headers.get('set-cookie');
					if (setCookie) this.cookies = setCookie.split(';')[0];
					return true;
				},
				{ maxAttempts: 2, delayMs: 1000 }
			);
			return success;
		} catch { return false; }
	}

	private async fetchTorrents(retried = false): Promise<TorrentInfo[]> {
		if (!(await this.login())) return [];
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/torrents/info?category=${encodeURIComponent(settings.QBIT_CATEGORY)}`, { headers: { Cookie: this.cookies } });
			if (resp.status === 403 && !retried) { this.cookies = ''; return this.fetchTorrents(true); }
			return resp.ok ? await resp.json() : [];
		} catch { return []; }
	}

	private async fetchTorrentComment(hash: string, retried = false): Promise<string | null> {
		if (!(await this.login())) return null;
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/torrents/properties?hash=${encodeURIComponent(hash)}`, { headers: { Cookie: this.cookies } });
			if (resp.status === 403 && !retried) { this.cookies = ''; return this.fetchTorrentComment(hash, true); }
			const props = await resp.json();
			return props.comment?.trim() || null;
		} catch { return null; }
	}

	async getActiveDownloads(): Promise<TorrentInfo[]> {
		return this.fetchTorrents();
	}

	async getTorrent(hash: string, retried = false): Promise<TorrentInfo | null> {
		if (!(await this.login())) return null;
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/torrents/info?hashes=${hash}`, { headers: { Cookie: this.cookies } });
			if (resp.status === 403 && !retried) { this.cookies = ''; return this.getTorrent(hash, true); }
			if (!resp.ok) return null;
			const torrents = await resp.json();
			return torrents.length > 0 ? torrents[0] : null;
		} catch { return null; }
	}

	async syncGames(): Promise<number> {
		const torrents = await this.fetchTorrents();
		if (!torrents.length) return 0;
		const syncStartTime = new Date().toISOString();
		
		let processedCount = 0;
		for (const torrent of torrents) {
			try {
				const allGames = db.select().from(games).all();
				const success = await this.processTorrent(torrent, allGames);
				if (success) processedCount++;
			} catch (error) {
				logger.error(`[qBit] Error processing torrent ${torrent.name || ''}: ${error}`);
			}
		}
		
		const currentGames = db.select().from(games).all();
		const gamesToUnlink = currentGames.filter(g => g.qbitSyncedAt && g.qbitSyncedAt < syncStartTime);
		for (const g of gamesToUnlink) {
			db.update(games).set({ qbitSyncedAt: null }).where(eq(games.id, g.id)).run();
			logger.info(`Unlinked '${g.title}'`);
		}
		return processedCount;
	}

	private async processTorrent(torrent: TorrentInfo, allGames: Array<typeof games.$inferSelect>): Promise<boolean> {
		const ts = torrent.added_on || torrent.completion_on || 0;
		if (ts <= 0) return false;

		const torrentDate = new Date(ts * 1000).toISOString();
		const rawName = torrent.name || '';
		const title = parseTorrentTitle(rawName);
		if (!title) return false;

		const detectedVersion = extractVersion(rawName);
		let sourceUrl = torrent.comment?.trim() || null;
		if (!sourceUrl || !sourceUrl.startsWith('http')) {
			sourceUrl = await this.fetchTorrentComment(torrent.hash);
		}

		const matchedGame = allGames.find((g) => {
			if (g.infoHash && g.infoHash === torrent.hash) return true;
			if (fuzzyMatchTitles(g.title, title)) return true;
			if (g.rawName) {
				const parsedRawName = parseTorrentTitle(g.rawName);
				if (parsedRawName && fuzzyMatchTitles(parsedRawName, title)) return true;
			}
			return false;
		});

		if (!matchedGame) {
			return await this.addNewGame(title, sanitizeSearchQuery(title), detectedVersion, torrentDate, sourceUrl, rawName, torrent.hash);
		} else {
			return await this.updateExistingGame(matchedGame, detectedVersion, torrentDate, sourceUrl, title, rawName, torrent.hash);
		}
	}

	private async addNewGame(title: string, searchQuery: string, version: string | null, torrentDate: string, sourceUrl: string | null, rawName: string, infoHash: string): Promise<boolean> {
		logger.info(`New game: ${title} (v${version || '?'})`);
		let coverUrl: string | null = null, steamAppId: number | null = null, igdbId: number | null = null, igdbName: string | undefined;

		if (isIgdbEnabled()) {
			try {
				const { getGameMetadata } = await import('./igdb.js');
				const { cleanGameTitle } = await import('./utils.js');
				const meta = await getGameMetadata(cleanGameTitle(title));
				if (meta) {
					coverUrl = meta.coverUrl || null;
					steamAppId = meta.steamAppId || null;
					igdbId = meta.igdbId || null;
					igdbName = meta.name;
				}
			} catch {}
		}

		db.insert(games).values({
			title: igdbName || title, searchQuery: igdbName || searchQuery,
			currentVersionDate: torrentDate, currentVersion: version,
			status: 'monitored', coverUrl, steamAppId, igdbId,
			sourceUrl, rawName, infoHash, isManual: false, qbitSyncedAt: new Date().toISOString()
		}).run();
		return true;
	}

	private async updateExistingGame(game: any, version: string | null, torrentDate: string, sourceUrl: string | null, newTitle: string, rawName: string, infoHash: string): Promise<boolean> {
		const updates: any = { qbitSyncedAt: new Date().toISOString(), isManual: false, rawName, infoHash };
		if (!game.igdbId && game.title !== newTitle) {
			logger.info(`Auto-rename: '${game.title}' -> '${newTitle}'`);
			updates.title = newTitle;
			updates.searchQuery = newTitle;
		}
		if (!game.sourceUrl && sourceUrl) updates.sourceUrl = sourceUrl;
		if (torrentDate > game.currentVersionDate) {
			updates.currentVersionDate = torrentDate;
			if (version) updates.currentVersion = version;
			if (sourceUrl) updates.sourceUrl = sourceUrl;
		} else if (!game.currentVersion && version) {
			updates.currentVersion = version;
		}
		db.update(games).set(updates).where(eq(games.id, game.id)).run();
		return true;
	}

	async addTorrent(magnetUrl: string): Promise<boolean> {
		if (!(await this.login())) return false;
		try {
			const success = await retryAsync(
				async () => {
					let body: any;
					const headers: any = { Cookie: this.cookies };

					if (magnetUrl.startsWith('http://') || magnetUrl.startsWith('https://')) {
						logger.debug(`[qBit] Downloading .torrent from URL before sending to client...`);
						const response = await fetch(magnetUrl);
						if (!response.ok) {
							logger.error(`[qBit] Failed to download .torrent file: ${response.status} ${response.statusText}`);
							return false;
						}
						const buffer = await response.arrayBuffer();
						const formData = new FormData();
						formData.append('torrents', new Blob([buffer]), 'download.torrent');
						formData.append('category', settings.QBIT_CATEGORY);
						body = formData;
					} else {
						headers['Content-Type'] = 'application/x-www-form-urlencoded';
						body = new URLSearchParams({ urls: magnetUrl, category: settings.QBIT_CATEGORY });
					}

					const resp = await fetch(`${this.baseUrl}/api/v2/torrents/add`, {
						method: 'POST',
						headers,
						body
					});
					return resp.ok;
				},
				{ maxAttempts: 2, delayMs: 1000 }
			);
			return success;
		} catch (error) {
			logger.error(`[qBit] Error adding torrent: ${error}`);
			return false;
		}
	}

	async removeTorrent(hash: string, deleteFiles = false, retried = false): Promise<boolean> {
		if (!hash || !(await this.login())) return false;
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/torrents/delete`, {
				method: 'POST',
				headers: { 'Content-Type': 'application/x-www-form-urlencoded', Cookie: this.cookies },
				body: new URLSearchParams({ hashes: hash, deleteFiles: deleteFiles ? 'true' : 'false' })
			});
			if (resp.status === 403 && !retried) { this.cookies = ''; return this.removeTorrent(hash, deleteFiles, true); }
			return resp.ok;
		} catch { return false; }
	}

	async pauseTorrent(hash: string, retried = false): Promise<boolean> {
		if (!hash || !(await this.login())) return false;
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/torrents/stop`, {
				method: 'POST',
				headers: { 'Content-Type': 'application/x-www-form-urlencoded', Cookie: this.cookies },
				body: new URLSearchParams({ hashes: hash })
			});
			if (resp.status === 403 && !retried) { this.cookies = ''; return this.pauseTorrent(hash, true); }
			return resp.ok;
		} catch { return false; }
	}

	async resumeTorrent(hash: string, retried = false): Promise<boolean> {
		if (!hash || !(await this.login())) return false;
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/torrents/start`, {
				method: 'POST',
				headers: { 'Content-Type': 'application/x-www-form-urlencoded', Cookie: this.cookies },
				body: new URLSearchParams({ hashes: hash })
			});
			if (resp.status === 403 && !retried) { this.cookies = ''; return this.resumeTorrent(hash, true); }
			return resp.ok;
		} catch { return false; }
	}

	async reannounceTorrent(hash: string, retried = false): Promise<boolean> {
		if (!hash || !(await this.login())) return false;
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/torrents/reannounce`, {
				method: 'POST',
				headers: { 'Content-Type': 'application/x-www-form-urlencoded', Cookie: this.cookies },
				body: new URLSearchParams({ hashes: hash })
			});
			if (resp.status === 403 && !retried) { this.cookies = ''; return this.reannounceTorrent(hash, true); }
			return resp.ok;
		} catch { return false; }
	}

	async recheckTorrent(hash: string, retried = false): Promise<boolean> {
		if (!hash || !(await this.login())) return false;
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/torrents/recheck`, {
				method: 'POST',
				headers: { 'Content-Type': 'application/x-www-form-urlencoded', Cookie: this.cookies },
				body: new URLSearchParams({ hashes: hash })
			});
			if (resp.status === 403 && !retried) { this.cookies = ''; return this.recheckTorrent(hash, true); }
			return resp.ok;
		} catch { return false; }
	}

	async getSpeedMode(): Promise<0 | 1 | null> {
		if (!(await this.login())) return null;
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/transfer/speedLimitsMode`, {
				headers: { Cookie: this.cookies }
			});
			if (!resp.ok) return null;
			const text = await resp.text();
			return parseInt(text.trim()) === 1 ? 1 : 0;
		} catch { return null; }
	}

	async toggleSpeedMode(retried = false): Promise<boolean> {
		if (!(await this.login())) return false;
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/transfer/toggleSpeedLimitsMode`, {
				method: 'POST',
				headers: { Cookie: this.cookies }
			});
			if (resp.status === 403 && !retried) { this.cookies = ''; return this.toggleSpeedMode(true); }
			return resp.ok;
		} catch { return false; }
	}

	async setTorrentDownloadLimit(hash: string, limitBytesPerSec: number, retried = false): Promise<boolean> {
		if (!hash || !(await this.login())) return false;
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/torrents/setDownloadLimit`, {
				method: 'POST',
				headers: { 'Content-Type': 'application/x-www-form-urlencoded', Cookie: this.cookies },
				body: new URLSearchParams({ hashes: hash, limit: limitBytesPerSec.toString() })
			});
			if (resp.status === 403 && !retried) { this.cookies = ''; return this.setTorrentDownloadLimit(hash, limitBytesPerSec, true); }
			return resp.ok;
		} catch { return false; }
	}

	async setTorrentUploadLimit(hash: string, limitBytesPerSec: number, retried = false): Promise<boolean> {
		if (!hash || !(await this.login())) return false;
		try {
			const resp = await fetch(`${this.baseUrl}/api/v2/torrents/setUploadLimit`, {
				method: 'POST',
				headers: { 'Content-Type': 'application/x-www-form-urlencoded', Cookie: this.cookies },
				body: new URLSearchParams({ hashes: hash, limit: limitBytesPerSec.toString() })
			});
			if (resp.status === 403 && !retried) { this.cookies = ''; return this.setTorrentUploadLimit(hash, limitBytesPerSec, true); }
			return resp.ok;
		} catch { return false; }
	}
}

export const qbitService = new QBitService();
